### PR TITLE
Movement

### PR DESCRIPTION
**Pull request by lsuyean**
## Overview

In this PR, I implement a new module Movement that handles the movement of the players can be animated or instantaneous. 

To turn on animations, set ```animate=True``` in ```robotouill-simulator.py```.

## Changes Made

- Added a new module ```Movement``` which controls the move action in robotouille. The class has a field ```animate```. If set to false, all movements are instantaneous. If set to true, each call to ```step()``` in ```state.py``` acts as 1 time step, and every tile moved by a player takes 1 time step. All other actions take 1 time step. 
- Added a new class ```Player``` that keeps tracks of player locations and movement paths. 
- Edited ```state.py``` to pass all move actions into the movement module
- Edited ```robotouille.json```, removed predicate ```vacant``` so that we allow for multiple players at one location. The path finding algorithm in ```Movement``` ensures that there are vacant spaces at a station to move to.
- Removed code from ```canvas.py``` that handled player movement. Now, it uses information from the ```Player``` class to know the position and direction of a player. 

## Test Coverage

Tested using all tests, all tests are working

### CODE REVIEW

File: backend/movement/movement.py [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+            player (Player): The player.
+            destination (tuple): The destination position.
+
+        Returns:
+            possible_destinations (List[tuple]): The possible destination positions.
+        """
+        possible_destinations = []
+        other_station_locations = self._get_station_locations()
+        player_locations = [(p.pos[0], p.pos[1]) for p in self.players.values() if p != player]
+        player_destinations = [p.path[-1] for p in self.players.values() if p != player and p.path != []]
```
Comment (chalo2000) [inline backend/movement/movement.py:119]: it is odd that the player maintains their path when the path is related to movement. the path a Player takes is not inherent to the Player but rather to Movement. if we forsee that the Player's path will be used by other modules, then it makes sense to leave `path` in Player; otherwise, it makes more sense for it to be maintained by Movement and the Player's position is updated by Movement (position makes sense to be in Player since it is likely to be used elsewhere, e.g. a module that checks if a Player is in range of something).

File: backend/movement/movement.py [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+        width, height = len(self.layout[0]), len(self.layout)
+        for i, j in [(0, 1), (0, -1), (1, 0), (-1, 0)]:
+            next_pos = (destination[0] + i, destination[1] + j)
+            if next_pos[0] < 0 or next_pos[0] >= width or next_pos[1] < 0 or next_pos[1] >= height:
```
Comment (chalo2000) [inline backend/movement/movement.py:123]: add a comment - 'Skipping - out of range'

File: backend/movement/movement.py [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+                continue
+            if next_pos in other_station_locations or next_pos in player_locations or next_pos in player_destinations:
```
Comment (chalo2000) [inline backend/movement/movement.py:125]: add a comment - 'Skipping - inside stations or players current/future locations'

File: backend/movement/movement.py [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+        return possible_destinations
+
+    def _get_player_path(self, player, destinations):
+        """
+        Gets the path for the player to move to the destination.
+
+        Args:
+            player (Player): The player.
+            destinations (List[tuple]): The possible destination positions
+                (accounts for other player locations and station locations when
```
Comment (chalo2000) [inline backend/movement/movement.py:137]: this parentheses information shouldn't be part of an argument's description but rather the longer function description to keep argument descriptions concise

File: backend/movement/movement.py [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+                continue
+            if current in visited:
+                continue
+            visited.add(current)
+            for i, j in [(0, 1), (0, -1), (1, 0), (-1, 0)]:
+                next_pos = (current[0] + i, current[1] + j)
+                new_path = path.copy()
+                new_path.append(next_pos)
+                queue.append((next_pos, new_path))
+        assert False, "No path found"
```
Comment (chalo2000) [inline backend/movement/movement.py:165]: can we rewrite the function so the path is returned at the end of the function? and we can add an assert just before the return statement to verify `current in destinations`. Also add this assertion error to the docstring.

File: backend/movement/movement.py [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+    
+    def move(self, state, player, destination, action, param_arg_dict):
+        """
+        Moves the player to the destination.
+
+        If animate mode is disabled, the player is immediately moved to the
+        destination. If animate mode is enabled, the player path is calculated
```
Comment (chalo2000) [inline backend/movement/movement.py:172]: as mentioned before - this should not be called `animate` mode.

File: backend/movement/movement.py [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+        and the player moves the first step in the path.
+
+        Args:
+            state (State): The state of the environment.
+            player (Object): The player to move.
```
Comment (chalo2000) [inline backend/movement/movement.py:177]: fix these types - there are classes corresponding to these arguments

File: backend/movement/movement.py [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+
+        Modifies:
+            player (Player): Modifies the path and pos fields of the player.
+        """
+        player_obj = self.players[player.name]
+        destination_pos = self.stations[destination.name]
+        possible_destinations = self._get_possible_destinations(player_obj, destination_pos)
+        # If player is already at the destination, perform move action immediately
+        if player_obj.pos in possible_destinations:
+            action.perform_action(state, param_arg_dict)
```
Comment (chalo2000) [inline backend/movement/movement.py:190]: i don't understand - if the player is at the possible destination what are we performing the action for? is this so it is set in the state so the appropriate predicates are set? if so, make the comment clearer

File: backend/movement/movement.py [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
+        else:
+            # Animate the movement by moving the player by on step in the path
+            player_obj.path = path
+            next_pos = player_obj.path.pop(0)
+            player_obj.direction = (next_pos[0] - player_obj.pos[0], next_pos[1] - player_obj.pos[1])
+            player_obj.pos = next_pos
+            # If player has reached the destination, perform the action
+            if player_obj.path == []:
+                action.perform_action(state, param_arg_dict)
+                destination = param_arg_dict["s2"]
```
Comment (chalo2000) [inline backend/movement/movement.py:209]: is it possible to combine this with the logic before returning if the player is already at the destination?

File: backend/movement/movement.py [hunk 10]
```diff
@@ -0,0 +1,249 @@
...
+                if player.path == []:
+                    player.motion = False
+                    player.action[0].perform_action(state, player.action[1])
+                    destination = player.action[1]["s2"]
+                    station_pos = self.stations[destination.name]
+                    player.direction = (station_pos[0] - player.pos[0], station_pos[1] - player.pos[1])
+                    player.action = None
+
+
+    def get_player(self, player_name):
```
Comment (chalo2000) [inline backend/movement/movement.py:239]: this doesn't seem to be used anywhere. but this would be a good class function to have in Player. see Player comments on how to maintain `self.players` in a clearer way

**Review by chalo2000 — CHANGES_REQUESTED** [review]
Added some comments on the existing code. Functionality looks good but architecturally there are better places to put variables to make the code clearer. In addition, this PR is still missing PyGame animations for movement so continue working on that.

File: backend/movement/movement.py [hunk 11]
```diff
@@ -0,0 +1,249 @@
+from backend.movement.player import Player
+
+class Movement(object):
+    """
+    This class is responsible for the movement in Robotouille.
+
+    This class keeps track of the positions of all objects in the environment, 
+    as well as the layout of the environment. 
+
+    It also handles the movement actions of Robotouille. Eacj call to the step
```
[the remark this answers is not in the document]
Comment (lsuyean) [reply backend/movement/movement.py:10]: fixed typo

File: backend/movement/movement.py [hunk 12]
```diff
@@ -0,0 +1,249 @@
...
+    function in the State class represents 1 time step. Every action other than 
+    move take 1 time step. When animate mode is disabled, move actions are
```
[the remark this answers is not in the document]
Comment (lsuyean) [reply backend/movement/movement.py]: Made this a new class Mode(Enum)

File: backend/movement/movement.py [hunk 13]
```diff
@@ -0,0 +1,249 @@
...
+        stations = {}
+        for station in environment_json["stations"]:
+            name = station["name"]
+            pos = (station["x"], station["y"])
+            stations[name] = pos
+        return stations
+    
+    def initialize(self, environment_json):
+        """
+        Initializes the movement.
```
[the remark this answers is not in the document]
Comment (lsuyean) [reply backend/movement/movement.py]: Player and station locations maintained by respective classes

File: robotouille/robotouille_simulator.py [hunk 14]
```diff
@@ -5,29 +5,73 @@
...
+        players = obs.get_players()
         actions = []
-        action, args = create_action_from_control(env, obs, obs.current_player, mousedown_events+keydown_events, renderer)
-        for player in obs.get_players():
-            if player == obs.current_player:
-                actions.append((action, args))
+        i = 0
+        while not done:
+            i += 1
+            if i%200000 != 0:
```
[the remark this answers is not in the document]
Comment (lsuyean) [reply robotouille/robotouille_simulator.py]: removed unnecessary if else code, now both movement modes share same code

File: backend/movement/player.py [hunk 15]
```diff
@@ -0,0 +1,29 @@
+class Player(object):
```
[the remark this answers is not in the document]
Comment (lsuyean) [reply backend/movement/player.py:1]: Class variable added to Player class

File: backend/movement/movement.py (hunk 1 above)
Comment (lsuyean) [reply backend/movement/movement.py]: Path now maintained by a class called PlayerData in movement.py

File: backend/movement/movement.py (hunk 2 above)
Comment (lsuyean) [reply backend/movement/movement.py:75]: added comment

File: backend/movement/movement.py (hunk 3 above)
Comment (lsuyean) [reply backend/movement/movement.py:78]: added comment

File: backend/movement/movement.py (hunk 4 above)
Comment (lsuyean) [reply backend/movement/movement.py:90]: Edited accordingly

File: backend/movement/movement.py (hunk 5 above)
Comment (lsuyean) [reply backend/movement/movement.py]: Path returned at end of function, assertion added, doc updated

File: backend/movement/movement.py (hunk 6 above)
Comment (lsuyean) [reply backend/movement/movement.py]: Updated

File: backend/movement/movement.py (hunk 7 above)
Comment (lsuyean) [reply backend/movement/movement.py]: Types fixed

File: backend/movement/movement.py (hunk 10 above)
Comment (lsuyean) [reply backend/movement/movement.py]: updated

File: renderer/configuration/robotouille_config.json [hunk 16]
```diff
@@ -1,12 +1,12 @@
...
-    "version": "1.0.0",
+    "version": "1.0.1",
 
     "player": {
         "robot": {
-            "front": "robot_front.png",
-            "back": "robot_back.png",
-            "left": "robot_left.png",
-            "right": "robot_right.png"
+            "front": ["robot_front_1.png", "robot_front_2.png", "robot_front_3.png", "robot_front_4.png"],
```
Comment (chalo2000) [inline renderer/configuration/robotouille_config.json:6]: This can be a temporary solution for now though `robot_front.png` and the other spritesheets are meant to be used directly by programmatically splitting it into rows and columns and choosing a grid cell based on the frame to display. This will require a new class and could be bundled in a separate PR for creating a dedicated animation class so let's make an issue for this.

File: robotouille/robotouille_simulator.py [hunk 17]
```diff
@@ -1,33 +1,51 @@
...
-            if player == obs.current_player:
-                actions.append((action, args))
-            else:
-                actions.append((None, None))
-        if not interactive and action is None:
+        player_obj = Player.get_player(obs.current_player.name)
+        no_action = True
+        
+        # If player is moving, do not allow any action; action will be None
+        if Movement.is_player_moving(player_obj.name):
```
Comment (chalo2000) [inline robotouille/robotouille_simulator.py:28]: no need for a `no_action` boolean, try this
```
# If player is moving, skip actions
if Movement.is_player_moving(player_obj.name):
  continue
action, args = create_action_from_control(env, obs, obs.current_player, mousedown_events+keydown_events, renderer)
if not interactive and action is None:
     # Retry for keyboard input
     continue
```
I believe you are trying to take environment steps because of the `clock.tick(20)` calls in movement but we fixed this by adding `env.render(...)` in this game loop which would just require the rendering to occur but the actual game timesteps to be fixed (assuming movement is always still 1 timestep)

**Review by chalo2000 — CHANGES_REQUESTED** [review]
Glad to see the robot moving and not teleporting for once. I left some comments that should be addressed and this PR seems to still rely on the `clock.tick(20)` calls which we fixed together last meeting by putting `env.render(...)` within the game loop. Please fix this and verify that the timesteps are the same as before (e.g. patties aren't cooked immediately upon moving but rather after 3 timesteps as in the configuration).

In addition to the last point, I would think about how we can make a timestep pass when the robot crosses a single tile (and how this logic will also work in multi-agent). We'll discuss this next meeting and gauge whether it is appropriate to include in this PR based on the amount of effort it'll be.

File: backend/movement/movement.py [hunk 18]
```diff
@@ -0,0 +1,224 @@
...
+            player_obj.direction = (destination_pos[0] - player_obj.pos[0], destination_pos[1] - player_obj.pos[1])
+            action.perform_action(state, param_arg_dict)
+        else:
+            # Animate the movement by moving the player by one step in the path
+            prev_pos = path[0]
+            next_pos = path[1]
+            data = PlayerData(path, 0)
+            Movement.player_movement_data[player.name] = data
+            player_obj.direction = (next_pos[0] - prev_pos[0], next_pos[1] - prev_pos[1])
+            clock.tick(20)
```
[the remark this answers is not in the document]
Comment (lsuyean) [reply backend/movement/movement.py:169]: removed ticks

File: backend/movement/player.py [hunk 19]
```diff
@@ -0,0 +1,52 @@
...
+        Builds the players in the environment.
+
+        Args:
+            environment_json (dict): The environment json.
+        """
+        for player in environment_json["players"]:
+            name = player["name"]
+            pos = (player["x"], player["y"])
+            direction = (player["direction"][0], player["direction"][1])
+            Player(name, pos, direction)
```
[the remark this answers is not in the document]
Comment (lsuyean) [reply backend/movement/player.py:39]: Moved building of Player.players to build_players

File: backend/movement/station.py [hunk 20]
```diff
@@ -0,0 +1,45 @@
...
+        """
+        Builds the stations in the environment.
+
+        Args:
+            environment_json (dict): The environment json.
+        """
+        for station in environment_json["stations"]:
+            name = station["name"]
+            pos = (station["x"], station["y"])
+            Station(name, pos)
```
[the remark this answers is not in the document]
Comment (lsuyean) [reply backend/movement/station.py:33]: Moved building of Staion.stations to build_stations

File: backend/state.py [hunk 21]
```diff
@@ -413,19 +420,28 @@ def step(self, actions):
...
         """
+        self.movement.step(self, clock)
+
         for action, param_arg_dict in actions:
             if not action:
                 continue
             assert action.is_valid(self, param_arg_dict)
-            self = action.perform_action(self, param_arg_dict)
+            # If the action is a movement action, use the movement module
+            if action.name == "move":
```
[the remark this answers is not in the document]
Comment (lsuyean) [reply backend/state.py:430]: I tried separating the logic for the step functions, so that the movement handles the logic. Currently, movement.step i still called within state.step, but movement.step handles all movement related functions and state.step handles everything else

File: robotouille/robotouille_simulator.py (hunk 17 above)
Comment (lsuyean) [reply robotouille/robotouille_simulator.py:28]: Issue made

File: renderer/configuration/robotouille_config.json (hunk 16 above)
Comment (lsuyean) [reply renderer/configuration/robotouille_config.json:6]: Issue made

File: robotouille/robotouille_simulator.py (hunk 17 above)
Comment (lsuyean) [reply robotouille/robotouille_simulator.py:28]: This code didnt work because currently we need to pass in a None action if a player is moving. This none action needs to be handled differently depending if the player is moving. If they are moving, we do not wait for another keyboard input and we step the environment. Else if they are moving and action is None, we retry for keyboard input.

File: backend/movement/movement.py [hunk 22]
```diff
@@ -0,0 +1,254 @@
+from backend.movement.player import Player
+from backend.movement.station import Station
+from enum import Enum
+
+class MetaData(object):
+    def __init__(self, path, time):
```
Comment (chalo2000) [inline backend/movement/movement.py:6]: a docstring including information on what path and time represents would be helpful

File: backend/movement/player.py [hunk 23]
```diff
@@ -0,0 +1,52 @@
...
+
+        Args:
+            environment_json (dict): The environment json.
+        """
+        for player in environment_json["players"]:
+            name = player["name"]
+            pos = (player["x"], player["y"])
+            direction = (player["direction"][0], player["direction"][1])
+            player_obj = Player(name, pos, direction)
+            Player.players[name] = player_obj
```
Comment (chalo2000) [inline backend/movement/player.py:39]: this should be done in `__init__` so that anyone that creates a Player without the `build_players` function doesn't have issues with `get_player(player_name)`.

File: backend/movement/station.py [hunk 24]
```diff
@@ -0,0 +1,45 @@
...
+        Builds the stations in the environment.
+
+        Args:
+            environment_json (dict): The environment json.
+        """
+        for station in environment_json["stations"]:
+            name = station["name"]
+            pos = (station["x"], station["y"])
+            station_obj = Station(name, pos)
+            Station.stations[name] = station_obj
```
Comment (chalo2000) [inline backend/movement/station.py:33]: ditto to Player comment

File: robotouille/robotouille_simulator.py [hunk 25]
```diff
@@ -1,33 +1,52 @@
...
 from utils.robotouille_input import create_action_from_control
 from robotouille.robotouille_env import create_robotouille_env
+from backend.movement.player import Player
+from backend.movement.movement import Movement, Mode
 
 
 def simulator(environment_name: str, seed: int=42, noisy_randomization: bool=False):
     # Your code for robotouille goes here
-    env, json, renderer = create_robotouille_env(environment_name, seed, noisy_randomization)
+    movement_mode = Mode.TRAVERSE
```
Comment (chalo2000) [inline robotouille/robotouille_simulator.py:11]: make this an argument passed in my command line and make the default mode `IMMEDIATE` so that we can tune the movement later to the cooking time when the game is in ready state.

**Review by chalo2000 — APPROVED** [review]
small changes to address but once done this LGTM.

File: backend/movement/movement.py (hunk 22 above)
Comment (lsuyean) [reply backend/movement/movement.py:6]: Added documentation

File: backend/movement/player.py (hunk 23 above)
Comment (lsuyean) [reply backend/movement/player.py:39]: Moved function to __init__

File: backend/movement/station.py (hunk 24 above)
Comment (lsuyean) [reply backend/movement/station.py:33]: Moved logit to __init__

File: robotouille/robotouille_simulator.py (hunk 25 above)
Comment (lsuyean) [reply robotouille/robotouille_simulator.py:11]: Edited accordingly
